### PR TITLE
AST: Introduce `MethodElement#getOverriddenMethods` to get all the methods overridden by this method

### LINKED ISSUE #10338: AST: Provide a way to get all overridden methods
### Feature description

Currently, we don't support the inheritance of the generic arguments across all inherited methods. I don't think there is an easy way to implement it in Micronaut 4. The only way is to traverse the generic arguments manually https://github.com/micronaut-projects/micronaut-validation/pull/228 but for that, we need a new API that will return those methods and doesn't slow the performance of the annotation processor.

### PR DESCRIPTION
**Pull request by dstepanov**
Reusing some of the logic from `EnclosedElementsQuery` to calculate the overridden methods for the annotation inheritance.

Fixes https://github.com/micronaut-projects/micronaut-core/issues/10338

### CODE REVIEW

File: inject-groovy/src/main/groovy/io/micronaut/ast/groovy/GroovyNativeElementHelper.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    private boolean isAssignable(ClassNode c1, ClassNode c2) {
+        if (c1.equals(c2)) {
+            return true;
+        }
+//        if (subDeclaringClass.isInterface()) {
+//            if (!superDeclaringClass.isInterface()) {
+//                return false;
+//            }
+//            return subDeclaringClass.declaresInterface(superDeclaringClass);
+//        }
```
Comment (graemerocher) [inline inject-groovy/src/main/groovy/io/micronaut/ast/groovy/GroovyNativeElementHelper.java]: remove this commented out code

File: inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java [hunk 2]
```diff
@@ -1,345 +0,0 @@
...
-import javax.lang.model.util.Types;
-import java.util.*;
-
-/**
- * Utility methods for dealing with generic type signatures.
- *
- * @author Graeme Rocher
- */
-@Internal
-public class GenericUtils {
```
Comment (graemerocher) [inline inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java:38]: are we sure this is not used by other modules?

File: inject-java/src/main/java/io/micronaut/annotation/processing/JavaNativeElementsHelper.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.List;
+
+/**
+ * The Java native element helper.
+ *
+ * @author Denis Stepanov
+ * @since 4.3.0
+ */
+@Internal
+public final class JavaNativeElementsHelper extends NativeElementsHelper<TypeElement, ExecutableElement> {
```
Comment (graemerocher) [inline inject-java/src/main/java/io/micronaut/annotation/processing/JavaNativeElementsHelper.java:42]: does it need to be public?

**Review by graemerocher — APPROVED** [review]

File: inject-java/src/main/java/io/micronaut/annotation/processing/JavaNativeElementsHelper.java (hunk 3 above)
Comment (dstepanov) [reply inject-java/src/main/java/io/micronaut/annotation/processing/JavaNativeElementsHelper.java:42]: It's used from different packages

**Comment by dstepanov** [comment]
@Spikhalskiy Would you mind testing the performance of this change in your project?

**Comment by Spikhalskiy** [comment]
@dstepanov It cuts kaptKotlin stage from ~20 sec to ~16-17 sec for us, so 15-20% improvement. Great. Thank you so much for all this build performance work. It greatly improves the experience with Micronaut and we are looking forward to 4.3.0 release.

**Comment by dstepanov** [comment]
@Spikhalskiy Thanks! I hope we will not lose much of the perf by implementing generic annotations inheritance that is missing now https://github.com/micronaut-projects/micronaut-validation/pull/228

**Comment by Spikhalskiy** [comment]
@dstepanov feel free to tag me when the fix for #9984 (or a localized fix for micronaut-validation) is merged, I will report how it looks on our end.

**Pull request merged**